pkg/controller/multiarchbuildconfig/mirror.go

```diff
@@ -0,0 +1,282 @@
+const (
+	// Condition types
+	MirrorImageManifestDone string = "ImageMirrorDone"
+
+	// Condition reasons
+	ImageMirrorErrorReason   string = "ImageMirrorError"
+	ImageMirrorSuccessReason string = "ImageMirrorSuccess"
+)
+
+type podState int
+
+const (
+	succeeded podState = iota
+	failed
+	undefined
+)
+
+// Watch for pods owned by some mabc and enqueue a request for the owner. It means that every
+// time a pod changes the main mabc reconciler is triggered.
+func watchPods(mgr manager.Manager, c controller.Controller) error {
+	if err := c.Watch(source.Kind(mgr.GetCache(), &corev1.Pod{}),
+		handler.EnqueueRequestForOwner(mgr.GetScheme(),
+			mgr.GetRESTMapper(),
+			&v1.MultiArchBuildConfig{},
+			handler.OnlyControllerOwner()),
+		predicate.Funcs{
+			CreateFunc:  func(e event.CreateEvent) bool { return isMABCMirrorPod(e.Object) },
+			DeleteFunc:  func(e event.DeleteEvent) bool { return isMABCMirrorPod(e.Object) },
+			UpdateFunc:  func(e event.UpdateEvent) bool { return isMABCMirrorPod(e.ObjectNew) },
+			GenericFunc: func(e event.GenericEvent) bool { return isMABCMirrorPod(e.Object) },
+		}); err != nil {
+		return fmt.Errorf("failed to watch %T pods: %w", v1.MultiArchBuildConfig{}, err)
+	}
+	return nil
+}
+
+func isMABCMirrorPod(obj client.Object) bool {
+	// Check the label only. We know it's a pod as we're only watching for those, see `source.Kind(&corev1.Pod{})`
+	_, exists := obj.GetAnnotations()[v1.MultiArchBuildConfigNameLabel]
+	return exists
+}
+
+// handleMirrorImage pushes an image to the locations specified in .spec.output.to. The function requires an image
+// to have been already created and pushed to the local registry. If all requirements are met it then spawns a new
+// mirror pod and wait for the result.
+// TODO: delete the mirror pod after it has completed
+// TODO: improve message on condition when a mirror pod fails
+func (r *reconciler) handleMirrorImage(ctx context.Context, srcImage string, mabc *v1.MultiArchBuildConfig) (func(mabcToMutate *v1.MultiArchBuildConfig), error) {
+	if len(mabc.Spec.Output.To) == 0 {
+		return nil, nil
+	}
+
+	// It doesn't make sense to keep going if the image with a manifest
+	// hasn't been properly pushed yet
+	pushDone := getCondition(mabc, PushImageManifestDone)
+	if pushDone == nil || pushDone.Status != metav1.ConditionTrue {
+		return nil, nil
+	}
+
+	// If the condition exists then we assume the mirror pod has already run
+	// so we don't need to do anything else
+	if getCondition(mabc, MirrorImageManifestDone) != nil {
+		return nil, nil
+	}
+
+	name, err := mirrorPodName(mabc)
+	if err != nil {
+		return nil, err
+	}
+	podName := types.NamespacedName{
+		Namespace: mabc.Namespace,
+		Name:      name,
+	}
+
+	// Either the pod doesn't exist yet so it has to be created or it does and we
+	// are going to check its state. Anything in between is treated as an error.
+	mirrorPod := corev1.Pod{}
+	if err := r.client.Get(ctx, podName, &mirrorPod, &client.GetOptions{}); err != nil {
```

Comment:
We don't need to spawn a new pod to mirror. The controller can just run it directly.

Reply:
I've changed the implementation; it now mirrors running `oc image run` without making any new pods.